File: src/bloom/domain/zone.py
from datetime import datetime

from pydantic import BaseModel
from shapely import Point, Polygon

from typing import Union


class Zone(BaseModel):
    id: int | None = None
    category: str
    sub_category: Union[str, None] = None
    name: str
    geometry: Union[Polygon, None] = None
    centroid: Union[Point, None] = None
    json_data: Union[dict, None] = None
    created_at: Union[ datetime , None] = None
    updated_at: Union[ datetime , None] = None

File: frontend/types/zone.ts
import { VesselDetails } from "./vessel"

export enum ZoneCategory {
  AMP = "amp",
  FISHING_COASTAL_WATERS = "Fishing coastal waters (6-12 NM)",
  TERRITORIAL_SEAS = "Territorial seas",
}

export type ZoneVisitTimeDto = {
  zone_id: number
  zone_category: string
  zone_sub_category: string
  zone_name: string
  visiting_duration: string
}

export interface ZoneDetails {
  id: string
  created_at: string
  updated_at: string
  category: string
  sub_category: string
  name: string
  centroid: {
    type: "Point"
    coordinates: number[]
  }
}

export interface ZoneMetrics {
  zone: ZoneDetails
  total_visit_duration: string
}

export type ZoneVesselMetrics = {
  zone: ZoneDetails
  vessel: VesselDetails
  vessel_visiting_time_by_zone: string
}

export type VesselZoneMetrics = {
  vessel: VesselDetails
  zone: ZoneDetails
  vessel_visiting_time_by_zone: string
}

export type ZoneWithGeometry = ZoneDetails & {
  geometry: {
    type: string
    coordinates: GeoJSON.Position[][][]
  }
}
